code/iaas/ha/src/main/java/io/cattle/platform/ha/monitor/impl/PingInstancesMonitorImpl.java

```diff
@@ -79,25 +79,22 @@ public void computeInstanceActivateReply(Event event) {
             instanceCache.invalidate(agentId);
         }
     }
-
-    protected boolean different(long agentId, Set<String> knownInstances, Set<String> instances, boolean check) {
-        instances = new HashSet<String>(instances);
-        for (String instance : knownInstances) {
-            if (!instances.remove(instance)) {
-                if (check) {
+    
+    protected boolean handleUnreportedKnownInstances(long agentId, Map<String, String> knownInstances, 
+            ReportedInstances reportedInstances, boolean checkOnly) {
+
+        for (Map.Entry<String, String> knownInstance : knownInstances.entrySet()) {
+            if(!reportedInstances.byUuid.contains(knownInstance.getKey()) 
+                    && !reportedInstances.byExternalId.contains(knownInstance.getValue())) {
+                if (checkOnly) {
                     return true;
                 } else {
-                    restart(instance);
+                    // If this known instance was not reported, schedule a potential restart.
+                    restart(knownInstance.getKey());
                 }
             }
         }
 
-        if (!check && COMPLAIN.get()) {
```

Comment:
We should probably do all the additional code cleanup if we plan to permanently remove this (removing the 'ha.instance.state.report.unknown' property)